### DISCUSSION TITLE
How to generate Thread Beacon message in OpenThread Network Simulator?

### DISCUSSION THREAD

**Discussion by MartinsMihaeljans** [openthread/openthread#9197] [Q&A] [answered]
Hello,
I have two questions about proper Thread network simulation with OTNS:

1) I tried to add a commissioner and a joiner to the network but got an error 13 - Invalid state both when I tried to make a router or FED as a joiner. **Can a joiner or commissioner be simulated if I use OTNS with visualization?**

<img width="595" alt="image" src="https://github.com/openthread/openthread/assets/128288466/31a4d633-1123-434c-b910-0beac40b02f1">

2) When I look at captured network traffic in current.pcap I find no Thread protocol messages. I'm relatively new to this wireless protocol stack so I have no idea whether the Thread message info included in MLE messages is all I should see or not. **Should I be able to capture for example Thread  beacon messages as well?**

Here is where Thread is mentioned in my capture:

<img width="599" alt="image" src="https://github.com/openthread/openthread/assets/128288466/b9d1c7a3-77b0-4d51-8388-c81af1b3ba55">

And here is example image:

<img width="397" alt="image" src="https://github.com/openthread/openthread/assets/128288466/51c88b38-967b-47ee-88aa-1b08aabb5384">

Example image is taken from this site - https://openthread.io/guides/pyspinel/sniffer

I understand this is a long post, but I'd be grateful if someone could help with this.

**Comment by jwhui** [comment]
> I tried to add a commissioner and a joiner to the network but got an error 13 - Invalid state both when I tried to make a router or FED as a joiner. Can a joiner or commissioner be simulated if I use OTNS with visualization?

Did you enable the interface before starting joiner operation? See [src/cli/README_COMMISSIONING.md#start-joiner](https://github.com/openthread/openthread/blob/main/src/cli/README_COMMISSIONING.md#start-joiner) for an example.

> When I look at captured network traffic in current.pcap I find no Thread protocol messages. I'm relatively new to this wireless protocol stack so I have no idea whether the Thread message info included in MLE messages is all I should see or not.

MLE is one of the protocols included in the [Thread Specification](https://www.threadgroup.org/support#specifications).

> Should I be able to capture for example Thread beacon messages as well?

Thread does not define a "Beacon" message. Do you mean IEEE 802.15.4 Beacon messages? If so, Thread only transmits IEEE 802.15.4 Beacons in response to an IEEE 802.15.4 Beacon Request. See the [IEEE 802.15.4 Active Scan command](https://github.com/openthread/openthread/tree/main/src/cli#scan-channel).

**Reply by MartinsMihaeljans** [reply]
@jwhui thank you for joining the discussion. I tried your suggested example, but something is still not correct. I did managed to enter the start joiner command but it only gave back Done output and not Join success. When I look at the datasets they are equal. But when I check commissioner joiner table, it is empty like in the picture:

<img width="333" alt="image" src="https://github.com/openthread/openthread/assets/128288466/bdcaa4b9-a1a7-48aa-b171-d6f3b6fd576a">

Also, I can not understand why Thread is specified as a protocol but does not come up in any pcap in protocol hierarchy?
Here is another example online where Wireshark does interpret Thread as a protocol:
![image](https://github.com/openthread/openthread/assets/128288466/47044b9c-d9fb-4d94-8390-8c40187c8998)
This example is from this site - https://wiki.makerdiary.com/nrf52840-mdk-usb-dongle/thread-sniffer/

Could this simply be a different Thread version in use?

**Reply by jwhui** [reply]
> @jwhui thank you for joining the discussion. I tried your suggested example, but something is still not correct. I did managed to enter the start joiner command but it only gave back Done output and not Join success. When I look at the datasets they are equal. But when I check commissioner joiner table, it is empty like in the picture:

That is expected. The joiner entry is removed after successful commissioning. Did you try starting Thread service?
```
$ thread start
```

> Also, I can not understand why Thread is specified as a protocol but does not come up in any pcap in protocol hierarchy?
Here is another example online where Wireshark does interpret Thread as a protocol:

Thread is not one protocol, but a collection of protocols, including MLE.

**Comment by MartinsMihaeljans** [comment]
Now I'm really confused. I understand I need to provide the next device that whishes to join the Thread network with pre shared key in my example J01ENM, but how do I do that? Like when I add a new SED for example:

<img width="491" alt="image" src="https://github.com/openthread/openthread/assets/128288466/cff00047-c5a8-43cc-b4da-77e01b66d226">

Also, do I understand correctly that thread is just a simple reuse of already available fields in MLE? Something like MultipathTCP within TCP where MPTCP is just an addon of TCP?

**Reply by jwhui** [reply, marked as answer]
> Now I'm really confused. I understand I need to provide the next device that whishes to join the Thread network with pre shared key in my example J01ENM, but how do I do that? Like when I add a new SED for example:

When using OTNS, you typically do not need to use the Thread Commissioning flow when adding nodes. See the [OTNS codelab](https://openthread.io/codelabs/openthread-network-simulator) provides a tutorial.

> Also, do I understand correctly that thread is just a simple reuse of already available fields in MLE? Something like MultipathTCP within TCP where MPTCP is just an addon of TCP?

Thread defines the MLE protocol. See Chapter 4 of [Thread Specification](https://www.threadgroup.org/support#specifications).

**Reply by MartinsMihaeljans** [reply]
@jwhui Thanks again for your help. I looked at the specification of thread and I still have an additional question about the capture:

1) Wireshark has a wide range of thread fields but in my capture only fields from meshcop and nwd can be found. Could it be that the rest are simply depreciated? Or are they used in specific case scenarios?

<img width="508" alt="image" src="https://github.com/openthread/openthread/assets/128288466/89c60a76-98c9-48e6-bc24-05d6bf489fc2">

Here is the Wireshark fields info:
<img width="754" alt="image" src="https://github.com/openthread/openthread/assets/128288466/1780a757-4e23-46bf-89d4-833fe7b04761">
https://www.wireshark.org/docs/dfref/

**Reply by MartinsMihaeljans** [reply]
@jwhui Thank you for your support. I marked your answer as I realized that my last question probably was out of scope of the original discussion topic. Sorry, for that. If you could elaborate bit on the last question that would be extremally helpful too. Like give me some jumpstart that's bit more easier then Thread specification.

**Reply by jwhui** [reply]
Some of those Thread-specific fields are in the payload of CoAP-TMF messages. Try to have packets sent to port 61631 decoded as `CoAP-TMF`.
